### ISSUE TITLE
How to access all features of the extension?

### ISSUE THREAD

**Issue by francisbarton** [seeba8/yet-another-pinboard-extension#22]
Hi, I am not sure if I am using the extension properly.

I am not sure what this means: "Also, allows you to view, open and edit your bookmarks in your browser." (from your README) - how do I access this feature? And is this any different from just visiting pinboard.in in a browser tab?

Also your keyboard shortcuts: r,t,u,n - as in the extension options - how do I use these? If I go to the omnibox and write "pin" + Tab then I get to search my bookmarks, but how do I use 't' for example to search tags only? It's not clear to me. Thanks for any help.

I had anticipated that the omnibox search would search my bookmarks in real time as dropdown options  in the box, but in fact the extension seems just to send the search query to a new tab with pinboard.in open. Not sure if this is what the extension is supposed to do?

In your README you talk about the options in the "Action Menu" but in Chrome I don't know what the Action Menu is. PLease could you clarify how to add the stated options?
Thank you.

**Comment by seeba8** [comment]
@francisbarton Hi! Thanks for trying the extension out. I'll try and clarify.

So, first of all, your first and last question I can answer in one: What I refer to there is the button in the browser bar, next to the url. If the button is not there, maybe try to reinstall it. I am not sure how to get it back in chrome. In firefox, you can go to the menu and click customize to get the icon back into the browser (action) bar.

![Action Button](https://www.dropbox.com/s/sg80o0pa74mcgxw/browseraction.png?raw=1)

When you click on that button, a list of all your bookmarks should appear. You can search them there, or in the normal browser address bar if you type "pin", a space, and then your search word.

For your second question about the shortcuts: That is probably worded badly in the extension, I will try to explain that better in my next patch. When you filter bookmarks in the address bar of the browser using `pin <searchword>`, you can also filter more precisely using for example:
* `pin t <searchword>` for bookmarks containing `<searchword>` in the tags
* `pin n <searchword>` for bookmarks containing `<searchword>` in the title
* etc.

![Search in address bar](https://www.dropbox.com/s/ysqtgekoueyu0re/omnibar_suggestions.png?raw=1)

I mostly use the popup when clicking on the button to search through my bookmarks, but the other functionality should work (if not, please do not hesitate to inform me!)

I hope that clarifies it a bit, please feel free to ask any more questions if something is unclear, if you need help, or if you want some additional feature.

**Comment by francisbarton** [comment]
Thank you Sebastian for your very patient reply.

I think the extension is not working here, (Chrome on Windows).
I have the button visible in the Chrome menu and I can use the links at the top of the popup to pin pages and go to pinboard website etc. But the search function does not return any results.

And when I use the omnibox to search, the "pin" shortcut works fine, but if I follow it with, say, "t politics" it just does a search for that exactly - it does not do a search for "politics" restricted to searching tags only.

Sorry!

**Comment by seeba8** [comment]
Hi Francis,

Sorry that it is not working. It sounds stupid to even say that, but it works for me (windows 10/chrome and firefox). I am sure there is a bug somewhere in my code. 

If you don't enter anything in the filter box, does it show your bookmarks in the popup?
If no, maybe try to go to the options, clear the api key, and re-set it, and maybe try to force-reload the key.
If that does not work, or if the error is something else, hmmm.

What you could do is go to `chrome://extensions`, check `Developer Mode` at  the top of the page, scroll to the extension, click on "Inspect views: background page" and check the developer console. If there are errors, paste them to me. If not, maybe then try to clear the key again and re-set / force reload it while the console is open. Then at least, there should be some errors.
Or you could check in the network tab of the console if it sends any requests to the pinboard API. It should do that when you click force-reload or when you re-set the key.

One more thing: if it still does not work, try restarting chrome. There might be an issue with my timeout-doubler, now that I think about it. What it does is: If a connection did not work, it doubles the time until the next try. So maybe, it did not work a few times and now the next attempt is somwhere next year.

When I am back home, I will check that code and maybe add a maximum delay of, say, 10 minutes.

I am sorry again that you have these issues.

**Comment by francisbarton** [comment]
Yes, the pop-up box is empty for me. I didn't realise it was not supposed
to be! I'll try starting again with the API key (I've tried force reload
but no change) and if that doesn't work I'll do some bug hunting!

**Comment by seeba8** [comment]
No no, it's supposed to look like this: 
![Pinboard screenshot](https://www.dropbox.com/s/p1dazyqax5oeqxu/pinboard-screenshot.png?raw=1)

My assumption then is that there are some problems when someone first installs it. That is strange.
Maybe I need to try and get a VM to test that.

**Comment by darpas70** [comment]
Dear Sebastian,
I've been using the Pinboard.in service for a while and now I have started to use your extension. I appreciate it very much since it provides me with a feature which is very useful for me: the sidebar in Firefox with my bookmarks saved in Pinboard.in. Anyway I am also encountering the same problem reported by user francisbarton in this thread, in fact also for me the bookmark search based on keyboard shortcuts: r,t,u,n does not work.
The problem is always present in Firefox (in sidebar, address bar and popup window). 
In Chrome I have observed a different behaviour, in fact searching in the address bar with the syntax "pin <string>" it correctly filters the results in real-time in the drop-down menu and it works also using the keyboard shortcuts (e.g. "pin t <string>") but not for all bookmarks (this is the most strange behaviour which also occurs sometimes in plain searches without keyboard shortcut). This behaviour is different from Firefox where using "pin <string>" in the address bar never returns real-time results in the drop-down menu, just after pressing "enter" open a pinboard.in page with filtered results.

For the rest the extension always works correctly and the bookmarks are correctly accessed and filtered by sidebar and popup window but not based on keyboard shortcut. Is there a possible fix to this problem? I'm fully available to support the debug if it may help

Thank you for your attention
Bye

**Comment by seeba8** [comment]
Hi @darpas70 
Thanks for the feedback. There may be a bug regarding the case sensitivity in my code. I think the search may only work if you type lowercase (since it converts the bookmarks to lowercase, but not the search term). 

The search in the popup and in the sidebar don't handle the prefixes (yet). That may be something I should implement. If you're up for some typescript development, I of course accept pull requests :)

> not for all bookmarks (this is the most strange behaviour which also occurs sometimes in plain searches without keyboard shortcut)

If you can give me some examples so that I can reproduce that, that would be amazing. I would need the search term, the expected resulting bookmark (including title, url, description, tags) so that I can do the same thing on my end. 

> Firefox where using "pin " in the address bar never returns real-time results in the drop-down menu

Really? That is working for me. Could you also give me an example (maybe including screenshot) for that so that I can try to reproduce the problem?
![image](https://user-images.githubusercontent.com/8963663/84577238-e5d9cd00-adba-11ea-933b-4375c3ccfa3e.png)

**Comment by seeba8** [comment]
Also, one thing to keep in mind is that the search in the address bar can only return 6 results (5 proper ones plus the default "search on pinboard" one. I may be able to fix it in a way that it will at least return 6, but more than that is not possible as per https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/omnibox/onInputChanged)

**Comment by darpas70** [comment]
Hello Sebastian,
thank you for your reply.
> The search in the popup and in the sidebar don't handle the prefixes (yet). That may be something I should implement. If you're up for some typescript development, I of course accept pull requests :)

I think that this is the main misunderstanding on my side. I was convinced that prefixes would work also in the sidebar which is my preferred way to access my bookmarks list. I will propose a pull request and hope that you have the possibility to implement. Thank you very much in advance :-)

For what concerns the other problems that I reported, subsequently to disinstalling and reinstalling the extension in Firefox I was unable to reproduce them and the search in the address bar seems now to work properly including prefixes. This is very strange and I do not have an explaination for this.

I thank you again for your support and will raise a pull request for the prefixes search in sidebar to be implemented

Best regards

**Comment by darpas70** [comment]
Hello Sebastian,
for what concerns the strange behaviour in Chrome I confirm it.
> 
> > not for all bookmarks (this is the most strange behaviour which also occurs sometimes in plain searches without keyboard shortcut)
> 
> If you can give me some examples so that I can reproduce that, that would be amazing. I would need the search term, the expected resulting bookmark (including title, url, description, tags) so that I can do the same thing on my end.
> 

This is a screenshot reporting my bookmarks tagged as "Halo"

<img width="848" alt="halo-pin" src="https://user-images.githubusercontent.com/66645661/85195416-9b80b080-b2d2-11ea-95f9-112171af8941.png">

and this is the address bar search working correctly:

<img width="960" alt="halo-address" src="https://user-images.githubusercontent.com/66645661/85195424-b3583480-b2d2-11ea-90bd-4c1c2ed99187.png">

On the other hand, these are my bookmarks tagged as "Ristoranti"

<img width="780" alt="ristoranti-pin" src="https://user-images.githubusercontent.com/66645661/85195437-cd921280-b2d2-11ea-8583-01110cb25210.png">

but the address bar search does not return any result for this tag (note that I have used lowercase anyway)

<img width="960" alt="ristoranti-address" src="https://user-images.githubusercontent.com/66645661/85195446-e4386980-b2d2-11ea-8a8e-54f8b035f742.png">

I hope that this can help to troubleshoot the problem

Thank you for your support

Best regards
